### PR TITLE
Support inlay hints

### PR DESCRIPTION
**Pull request by 333fred**
Closes https://github.com/OmniSharp/omnisharp-roslyn/issues/1932. Implements the new inlay hint api for vscode, now that it's stable and in LSP. Tests still need to be written.

### CODE REVIEW

**Comment by 333fred** [comment]
This will need https://github.com/OmniSharp/omnisharp-roslyn/pull/2357 to be merged, and I still need to write some integration tests covering the new feature. Some screenshots:

image
image

**Comment by somedeveloperhappy** [comment]
I just wanted to come here and thank you guys for all your efforts. the inlay hints was one of the main reasons I still have Rider installed on my machine

File: src/omnisharp/options.ts [hunk 1]
```diff
@@ -35,6 +35,18 @@ export class Options {
         public enableImportCompletion: boolean,
         public enableAsyncCompletion: boolean,
         public useSemanticHighlighting: boolean,
+        public enableForParameters: boolean,
```
Comment (JoeRobich) [inline src/omnisharp/options.ts:38]: Would you be open to renaming these all with an `inlayHints` prefix? ala `inlayHintsEnabledForParameters`

File: test/unitTests/Fakes/FakeOptions.ts [hunk 2]
```diff
@@ -6,5 +6,52 @@
...
+        /* forImplicitVariableTypes */false,
+        /* forLambdaParameterTypes */false,
+        /* forImplicitObjectCreation */false,
+        /* razorPluginPath */"",
+        /* defaultLaunchSolution */"",
+        /* monoPath */"",
+        /* dotnetPath */"",
+        /* excludePaths */null,
+        /* maxProjectFileCountForDiagnosticAnalysis */null,
+        /* testRunSettings */"");
```
Comment (JoeRobich) [inline test/unitTests/Fakes/FakeOptions.ts:56]: Thanks for this. It is both beautiful and scary to look at.

**Review by JoeRobich — COMMENTED** [review]
Did you test this with a Razor scenario?

Do we need to filter out Razor generated documents? 
https://github.com/OmniSharp/omnisharp-vscode/blob/84f1d5ab16718c2e0ad91e732770515500a8113e/src/features/virtualDocumentTracker.ts#L25-L39

Or handle mapping location information to Razor documents by passing it through the LanguageMiddleware?
https://github.com/OmniSharp/omnisharp-vscode/blob/84f1d5ab16718c2e0ad91e732770515500a8113e/src/features/referenceProvider.ts#L25-L27

**Comment by 333fred** [comment]
I think we should probably avoid razor documents for now, until @NTaylorMullen has a chance to test it out and make sure that everything is working as expected.

File: test/integrationTests/inlayHints.integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        csharpConfig.update('inlayHints.types.forImplicitObjectCreation', true);
+
+        await restartOmniSharpServer();
+        const activation = await activateCSharpExtension();
+        await testAssetWorkspace.restoreAndWait(activation);
+
+        const fileName = 'inlayHints.cs';
+        const projectDirectory = testAssetWorkspace.projects[0].projectDirectoryPath;
+        const filePath = path.join(projectDirectory, fileName);
+        fileUri = vscode.Uri.file(filePath);
```
Comment (JoeRobich) [inline test/integrationTests/inlayHints.integration.test.ts:51]: Might be best to open the file during setup and wait for events to settle before running the test. This is what we do in the diagnostics tests.

```suggestion
        await restartOmniSharpServer();
        const activation = await activateCSharpExtension();
        await testAssetWorkspace.restore();

        const fileName = 'inlayHints.cs';
        const projectDirectory = testAssetWorkspace.projects[0].projectDirectoryPath;
        const filePath = path.join(projectDirectory, fileName);
        fileUri = vscode.Uri.file(filePath);
        
        await vscode.commands.executeCommand("vscode.open", fileUri);
        await testAssetWorkspace.waitForIdle(activation.eventStream);
```

File: test/integrationTests/inlayHints.integration.test.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    });
+
+    suiteTeardown(async () => {
+        await testAssetWorkspace.cleanupWorkspace();
+    });
+
+    test("Hints retrieved for region", async () => {
+        await vscode.commands.executeCommand("vscode.open", fileUri);
```
Comment (JoeRobich) [inline test/integrationTests/inlayHints.integration.test.ts:59]: If you take the above suggestion, there is no need to open the file in the test.
```suggestion
```

**Review by JoeRobich — COMMENTED** [review]
Suggestion on how to make the test more consistent

**Comment by somedeveloperhappy** [comment]
hey. I really need this feature even if it's in beta. can someone help me on how to use this branch in vscode? I could be helpful with my bug reports later :p

**Comment by 333fred** [comment]
@JoeRobich we'll see, but locally the test doesn't even require that the file is open at all. Something else is happening.

**Comment by 333fred** [comment]
> hey. I really need this feature even if it's in beta. can someone help me on how to use this branch in vscode? I could be helpful with my bug reports later :p

Check out the branch and build it.

**Comment by somedeveloperhappy** [comment]
> > hey. I really need this feature even if it's in beta. can someone help me on how to use this branch in vscode? I could be helpful with my bug reports later :p
> 
> Check out the branch and build it.

yes, trying. but my npm gives out too many errors :(
is there some magic command that installs all needed npm packages?
+ also if you don't have time, I'll still be very glad if you at least tell me what to search/google about that

File: test/integrationTests/inlayHints.integration.test.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        csharpConfig.update('inlayHints.parameters.forObjectCreationParameters', true);
+        csharpConfig.update('inlayHints.parameters.forIndexerParameters', true);
+        csharpConfig.update('inlayHints.parameters.forOtherParameters', true);
+        csharpConfig.update('inlayHints.parameters.suppressForParametersThatDifferOnlyBySuffix', true);
+        csharpConfig.update('inlayHints.parameters.suppressForParametersThatMatchMethodIntent', true);
+        csharpConfig.update('inlayHints.parameters.suppressForParametersThatMatchArgumentName', true);
+        csharpConfig.update('inlayHints.types.enabled', true);
+        csharpConfig.update('inlayHints.types.forImplicitVariableTypes', true);
+        csharpConfig.update('inlayHints.types.forLambdaParameterTypes', true);
+        csharpConfig.update('inlayHints.types.forImplicitObjectCreation', true);
```
Comment (JoeRobich) [inline test/integrationTests/inlayHints.integration.test.ts:42]: Noticed in the CodeLens test that `.update()` is thenable and can be awaited.

```suggestion
        const editorConfig = vscode.workspace.getConfiguration('editor');
        await editorConfig.update('inlayHints.enabled', true);

        const csharpConfig = vscode.workspace.getConfiguration('csharp');
        await csharpConfig.update('inlayHints.parameters.enabled', true);
        await csharpConfig.update('inlayHints.parameters.forLiteralParameters', true);
        await csharpConfig.update('inlayHints.parameters.forObjectCreationParameters', true);
        await csharpConfig.update('inlayHints.parameters.forIndexerParameters', true);
        await csharpConfig.update('inlayHints.parameters.forOtherParameters', true);
        await csharpConfig.update('inlayHints.parameters.suppressForParametersThatDifferOnlyBySuffix', true);
        await csharpConfig.update('inlayHints.parameters.suppressForParametersThatMatchMethodIntent', true);
        await csharpConfig.update('inlayHints.parameters.suppressForParametersThatMatchArgumentName', true);
        await csharpConfig.update('inlayHints.types.enabled', true);
        await csharpConfig.update('inlayHints.types.forImplicitVariableTypes', true);
        await csharpConfig.update('inlayHints.types.forLambdaParameterTypes', true);
        await csharpConfig.update('inlayHints.types.forImplicitObjectCreation', true);
```

**Review by filipw — APPROVED** [review]

**Review by JoeRobich — APPROVED** [review]

**Comment by somedeveloperhappy** [comment]
Hey, umm , this is written "Merged" does it mean It's officially gonna go to the master branch now? Or just to the master of this fork?

Sorry I don't know much about GitHub

**Comment by 333fred** [comment]
This isn't a fork, it's the main repository. And yes, it has been merged to master and will be in the next release of omnisharp vscode.

**Comment by somedeveloperhappy** [comment]
> 

Thanks, yeah, figured that out by some research afterwards :D thanks. And the feature is really great! You guys are breaking the boundaries between vscode and Rider for dotnet! 

+ Though the new omnisharp loads suggestions after maybe a whole 1 second :( I'm assuming it's got something to do with the omnisharp itself and not the inlay hints, but just wanted to let you know

**Comment by JoeRobich** [comment]
@somedeveloperhappy Do you have either of these options enabled? Async completion should help with completion performance.
```json
    "omnisharp.enableAsyncCompletion": true,
    "omnisharp.enableImportCompletion": true,
```

**Pull request merged**